### PR TITLE
fix: add launchctl launcher for the agent

### PR DESCRIPTION
**Pull request by egorgripasov**
### Summary
🍎🚀🛠️

Added support for system daemon on macOS using launchctl for the `agent` commands in the `dxos devtools` CLI. Refactored the daemon management logic in the `dxos core agent` module and renamed the `phoenix` module to `launcher`. Improved the error handling and type safety of the `reset` and `stop` commands.

> _We are the launchers of the system daemon_
> _We control the process with the `launchctl` command_
> _We can start, stop, list and reset the agent_
> _We are the masters of the `execWithDaemon` method_

### Walkthrough
*  Refactored the daemon management logic to use launchctl on macOS and exported the `launcher` module instead of the `phoenix` module ([link](https://github.com/dxos/dxos/pull/4375/files?diff=unified&w=0#diff-7077a9101dd1e60023a49a63f848ad0642cf6286e53e00c99ea8fabba6fcdf22L8-R8), [link](https://github.com/dxos/dxos/pull/4375/files?diff=unified&w=0#diff-5bd5654aa245b5ebd558daefbff6e5baa47b5fc639ecc105e39b0a72c23c966aR5), [link](https://github.com/dxos/dxos/pull/4375/files?diff=unified&w=0#diff-67920a9063ba4edb9b551afa4434d62dc34f27494d1fddad00b9d62e91d92d21L1-R153), [link](https://github.com/dxos/dxos/pull/4375/files?diff=unified&w=0#diff-236f70408d4709ef20bf11e8a72a1395086af03e631fefae2bb32e556d6a54ddR1-R65), [link](https://github.com/dxos/dxos/pull/4375/files?diff=unified&w=0#diff-c8d119530cd3b793d516ca48eb5339ec2a0687659ae8382118e134be7381f280)).

### CODE REVIEW

File: packages/core/agent/src/launcher/launchctl-runner.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  async isRunning(service: string): Promise<boolean> {
+    try {
+      const { stdout } = await execPromise('launchctl list');
+      return stdout.includes(service);
+    } catch (err: any) {
+      throw new Error(`Failed to check if the service is running: ${err.message}`);
+    }
+  }
+
+  private async writeFile(filePath: string, content: string): Promise<void> {
```
Comment (mykola-vrmchk) [inline packages/core/agent/src/launcher/launchctl-runner.ts:54]: Is this method needed?

**Review by mykola-vrmchk — APPROVED** [review]

File: packages/core/agent/src/launcher/launchctl-daemon.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+import { LaunchctlRunner } from './launchctl-runner';
+import { Daemon, ProcessInfo, StartOptions, StopOptions } from '../daemon';
+import { CHECK_INTERVAL, DAEMON_STOP_TIMEOUT } from '../defs';
+import { AgentWaitTimeoutError } from '../errors';
+import { lockFilePath, removeLockFile, removeSocketFile, waitForAgentToStart } from '../util';
+
+/**
+ * Manager of system daemon processes using launchctl on macOS.
+ */
+export class LaunchctlDaemon implements Daemon {
```
Comment (mykola-vrmchk) [inline packages/core/agent/src/launcher/launchctl-daemon.ts:21]: Why do we separate classes LaunchctlDaemon and LaunchctlRunner? We could refactor LaunchctlDaemon to SystemDaemon, which will be less specific and accept different system-specific launchers.

File: packages/core/agent/src/launcher/system-daemon.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ * Manager of system daemon processes using system service manager.
+ */
+export class SystemDaemon implements Daemon {
+  private readonly _runner: Runner;
+
+  constructor(private readonly _rootDir: string) {
+    if (platform() === 'darwin') {
+      this._runner = new LaunchctlRunner();
+    } else {
+      throw new Error(`System daemon not implemented for ${platform()} yet.`);
```
Comment (richburdon) [inline packages/core/agent/src/launcher/system-daemon.ts:29]: remove "yet"

File: packages/core/agent/src/launcher/system-daemon.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+import { lockFilePath, removeLockFile, removeSocketFile, waitForAgentToStart } from '../util';
+
+/**
+ * Manager of system daemon processes using system service manager.
+ */
+export class SystemDaemon implements Daemon {
+  private readonly _runner: Runner;
+
+  constructor(private readonly _rootDir: string) {
+    if (platform() === 'darwin') {
```
Comment (richburdon) [inline packages/core/agent/src/launcher/system-daemon.ts:26]: generally try not to throw in constructors, unless via invariants.

File: packages/core/agent/src/launcher/system-daemon.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      this._runner = new LaunchctlRunner();
+    } else {
+      throw new Error(`System daemon not implemented for ${platform()} yet.`);
+    }
+
+    const dir = path.join(this._rootDir, 'profile');
```
Comment (richburdon) [inline packages/core/agent/src/launcher/system-daemon.ts:32]: extract and reuse const.

File: packages/devtools/cli/src/commands/reset/index.ts [hunk 6]
```diff
@@ -57,7 +57,15 @@ export default class Reset extends BaseCommand<typeof Reset> {
...
-      });
+      }, false);
+      // TODO(egorgripasov): Better compatibility with Phoenix daemon.
+      try {
+        await this.execWithDaemon(async (daemon) => {
+          if (await daemon.isRunning(this.flags.profile)) {
+            await daemon.stop(this.flags.profile, { force: this.flags.force });
+          }
+        }, true);
+      } catch (err) {}
```
Comment (richburdon) [inline packages/devtools/cli/src/commands/reset/index.ts:68]: Document silent error. or `// Ignore.` if expected.

File: packages/core/agent/src/launcher/runner/launchctl-runner.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  private async _getPlistPath(service: string): Promise<string> {
+    const agentsPath = `${os.homedir()}/Library/LaunchAgents`;
+    if (!existsSync(agentsPath)) {
+      await mkdir(agentsPath, { recursive: true });
+    }
+    return `${agentsPath}/${service}.plist`;
+  }
+
+  private _getServiceName(profile: string): string {
+    return `org.dxos.agent.${profile}`;
```
Comment (richburdon) [inline packages/core/agent/src/launcher/runner/launchctl-runner.ts:100]: Extract const. What standard are we using here? (Elsewhere dxos.org/service/xxx?)

Comment (egorgripasov) [reply packages/core/agent/src/launcher/runner/launchctl-runner.ts:100]: This is the macos specific format for services, see
https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html
or https://ss64.com/osx/launchctl.html

File: packages/devtools/cli/src/commands/reset/index.ts (hunk 6 above)
Comment (egorgripasov) [reply packages/devtools/cli/src/commands/reset/index.ts:68]: fixed in further PRs

**Pull request merged**